### PR TITLE
refactor: Convert SchedTest to Kotlin

### PR DESCRIPTION
**Pull request by david-allison**
Closes #11837

### CODE REVIEW

File: AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedTest.kt [hunk 1]
```diff
@@ -13,244 +13,190 @@
...
+import org.hamcrest.Matchers.*
+import org.junit.Ignore
+import org.junit.Test
+import org.junit.runner.RunWith
+import kotlin.test.*
+import kotlin.test.junit5.JUnit5Asserter.assertEquals
+
+@RustCleanup("Remove, or add KotlinCleanup")
+@RunWith(AndroidJUnit4::class)
+class SchedTest : RobolectricTest() {
```
Comment (lukstbit) [inline AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedTest.kt:59]: @KotlinCleanup("is -> equalTo")
@KotlinCleanup("fix ide lint issues")

File: AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedTest.kt [hunk 2]
```diff
@@ -13,244 +13,190 @@
...
+import com.ichi2.libanki.Consts.STARTING_FACTOR
+import com.ichi2.libanki.DecksTest.Companion.TEST_DECKS
+import com.ichi2.libanki.stats.Stats.Companion.SECONDS_PER_DAY
+import com.ichi2.libanki.utils.TimeManager
+import com.ichi2.testutils.AnkiAssert.checkRevIvl
+import com.ichi2.testutils.AnkiAssert.without_unicode_isolation
+import com.ichi2.testutils.MockTime
+import com.ichi2.testutils.MutableTime
+import com.ichi2.utils.JSONArray
+import net.ankiweb.rsdroid.BackendFactory.defaultLegacySchema
```
Comment (lukstbit) [inline AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedTest.kt:47]: Revert this import

File: AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedTest.kt [hunk 3]
```diff
@@ -13,244 +13,190 @@
...
 
-
-    @NonNull
-    private DeckDueTreeNode getCountsForDid(double didToFind) {
-        List<TreeNode<DeckDueTreeNode>> tree = getCol().getSched().deckDueTree();
-
-        for (TreeNode<DeckDueTreeNode> node : tree) {
-            if (node.getValue().getDid() == didToFind) {
-                return node.getValue();
+    private fun getCountsForDid(didToFind: Double): DeckDueTreeNode {
```
Comment (lukstbit) [inline AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedTest.kt:101]: @KotlinCleanup("simplify fun with firstOrNull and ?: ")

File: AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedTest.kt [hunk 4]
```diff
@@ -276,352 +222,350 @@ public class SchedTest extends RobolectricTest {
...
-        c.setType(CARD_TYPE_REV);
-        c.setQueue(QUEUE_TYPE_REV);
-        c.setDue(col.getSched().getToday());
-        c.setReps(1);
-        c.setIvl(1);
-        c.startTimer();
-        c.flush();
-        col.reset();
+        val c = note.cards()[0]
+        c.type = CARD_TYPE_REV
```
Comment (lukstbit) [inline AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedTest.kt:552]: @KotlinCleanup("use scope function")

File: AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedTest.kt [hunk 5]
```diff
@@ -657,396 +601,411 @@ public class SchedTest extends RobolectricTest {
...
-        assertThat(c.getDue(), is(greaterThanOrEqualTo(TimeManager.INSTANCE.getTime().intTime())));
-        assertEquals(QUEUE_TYPE_LRN, c.getQueue());
-        assertEquals(CARD_TYPE_REV, c.getType());
-        col.getSched().suspendCards(new long[] {c.getId()});
-        col.getSched().unsuspendCards(new long[] {c.getId()});
-        c.load();
-        assertEquals(QUEUE_TYPE_REV, c.getQueue());
-        assertEquals(CARD_TYPE_REV, c.getType());
-        assertEquals(1, c.getDue());
+        c.due = 0
```
Comment (lukstbit) [inline AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedTest.kt:738]: @KotlinCleanup("use apply")

File: AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedTest.kt [hunk 6]
```diff
@@ -657,396 +601,411 @@ public class SchedTest extends RobolectricTest {
...
-        c.setQueue(QUEUE_TYPE_REV);
-        c.setType(CARD_TYPE_REV);
+    @Throws(Exception::class)
+    fun test_cram() {
+        val col = getColV1()
+        val note = col.newNote()
+        note.setItem("Front", "one")
+        col.addNote(note)
+        var c = note.cards()[0]
+        c.ivl = 100
```
Comment (lukstbit) [inline AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedTest.kt:777]: @KotlinCleanup("use scope function")

File: AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedTest.kt [hunk 7]
```diff
@@ -657,396 +601,411 @@ public class SchedTest extends RobolectricTest {
...
+        c.load()
+        assertEquals(4, col.sched.answerButtons(c).toLong())
         // add a sibling so we can test minSpace, etc
-        Card c2 = c.clone();
-        c2.setId(0);
-        c2.setOrd(1);
-        c2.setDue(325);
-        c2.flush();
+        val c2 = c.clone()
+        c2.id = 0
```
Comment (lukstbit) [inline AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedTest.kt:873]: @KotlinCleanup("use apply")

File: AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedTest.kt [hunk 8]
```diff
@@ -657,396 +601,411 @@ public class SchedTest extends RobolectricTest {
...
-        col.getSched().rebuildDyn(did);
-        col.reset();
-        c = getCard();
-        assertEquals(600, col.getSched().nextIvl(c, BUTTON_ONE));
-        assertEquals(0, col.getSched().nextIvl(c, BUTTON_TWO));
-        assertEquals(0, col.getSched().nextIvl(c, BUTTON_THREE));
-        col.getSched().answerCard(c, BUTTON_TWO);
-        assertEquals(100, c.getIvl());
-        assertEquals(col.getSched().getToday() + 25, c.getDue());
+        c.ivl = 100
```
Comment (lukstbit) [inline AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedTest.kt:934]: @KotlinCleanup("use scope function")

File: AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedTest.kt [hunk 9]
```diff
@@ -1057,151 +1016,160 @@ public class SchedTest extends RobolectricTest {
...
-        c.flush();
-        col.reset();
-        assertEquals(new Counts(0, 0, 1), col.getSched().counts());
-        col.getSched().answerCard(getCard(), BUTTON_ONE);
-        assertEquals(new Counts(0, 1, 0), col.getSched().counts());
+        note = col.newNote()
+        note.setItem("Front", "three")
+        col.addNote(note)
+        val c = note.cards()[0]
+        c.type = CARD_TYPE_REV
```
Comment (lukstbit) [inline AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedTest.kt:1137]: @KotlinCleanup("use scope function")

File: AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedTest.kt [hunk 10]
```diff
@@ -1057,151 +1016,160 @@ public class SchedTest extends RobolectricTest {
...
-            c.setType(CARD_TYPE_REV);
-            c.setQueue(QUEUE_TYPE_REV);
-            c.setDue(0);
-            c.flush();
+        for (i in 0..4) {
+            val note = col.newNote()
+            note.setItem("Front", "num$i")
+            col.addNote(note)
+            val c = note.cards()[0]
+            c.type = CARD_TYPE_REV
```
Comment (lukstbit) [inline AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedTest.kt:1157]: @KotlinCleanup("use scope function")

File: AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedTest.kt [hunk 11]
```diff
@@ -1355,95 +1327,95 @@ public class SchedTest extends RobolectricTest {
...
-        col.reset();
-        assertEquals(new Counts(1, 0, 0), col.getSched().counts());
+    @Throws(Exception::class)
+    fun test_forgetV1() {
+        val col = getColV1()
+        val note = col.newNote()
+        note.setItem("Front", "one")
+        col.addNote(note)
+        val c = note.cards()[0]
+        c.queue = QUEUE_TYPE_REV
```
Comment (lukstbit) [inline AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedTest.kt:1338]: @KotlinCleanup("use scope function")

File: AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedTest.kt [hunk 12]
```diff
@@ -1355,95 +1327,95 @@ public class SchedTest extends RobolectricTest {
...
-        c.flush();
-        col.reset();
-        col.getSched().answerCard(c, BUTTON_ONE);
-        col.getSched()._cardConf(c).getJSONObject("lapse").put("delays", new JSONArray(new double[] {}));
-        col.getSched().answerCard(c, BUTTON_ONE);
+        val note = col.newNote()
+        note.setItem("Front", "one")
+        col.addNote(note)
+        val c = note.cards()[0]
+        c.type = CARD_TYPE_REV
```
Comment (lukstbit) [inline AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedTest.kt:1379]: @KotlinCleanup("use scope function")

File: AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedTest.kt [hunk 13]
```diff
@@ -1355,95 +1327,95 @@ public class SchedTest extends RobolectricTest {
...
-        assertEquals(25, c.getIvl());
+    @Throws(Exception::class)
+    fun test_failmultV1() {
+        val col = getColV1()
+        val note = col.newNote()
+        note.setItem("Front", "one")
+        note.setItem("Back", "two")
+        col.addNote(note)
+        var c = note.cards()[0]
+        c.type = CARD_TYPE_REV
```
Comment (lukstbit) [inline AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedTest.kt:1403]: @KotlinCleanup("use scope function")

**Review by lukstbit — CHANGES_REQUESTED** [review]
Looks ok, some small cleanups...

**Review by lukstbit — APPROVED** [review]

**Review by BrayanDSO — APPROVED** [review]

**Pull request merged**